File: abstractfactory/myabstractfactory.cpp
#include <iostream>

using namespace std;

class Fruit {
    public:
        virtual ~Fruit(){};
        virtual string Consume() const = 0;
};

class Banana: public Fruit {
    public:
        string Consume() const override {
            return "Banana: Peel and Eat";
        }
};

class Apple: public Fruit {
    public:
        string Consume() const override {
            return "Apple: Cut and Eat";
        }
};

class Vegetable {
    public:
        virtual ~Vegetable(){};
        virtual string CookAndConsume() const = 0;
        virtual string Serve() const = 0;
};

class Potato: public Vegetable {
    public:
        string CookAndConsume() const override {
            return "Potato: Fry and Eat";
        }

        string Serve() const override {
            return "Potato: In a bag";
        }
};

class Beans: public Vegetable {
    public:
        string CookAndConsume() const override {
            return "Beans: Boil and Eat";
        }

        string Serve() const override {
            return "Beans: In a bowl";
        }
};

class MyFactory {
    public:
        virtual ~MyFactory(){};
        virtual Fruit *ConsumeFruit() const = 0;
        virtual Vegetable *ConsumeVegetable() const = 0;
};

class FirstFactory: public MyFactory {
    public:
        Fruit *ConsumeFruit() const override {
            return new Banana();
        }
        Vegetable *ConsumeVegetable() const override {
            return new Potato();
        }
};

class SecondFactory: public MyFactory {
    public:
        Fruit *ConsumeFruit() const override {
            return new Apple();
        }
        Vegetable *ConsumeVegetable() const override {
            return new Beans();
        }
};

void Usage(const MyFactory &factory) {
    const Fruit *fruit = factory.ConsumeFruit();
    const Vegetable *vegetable = factory.ConsumeVegetable();

    cout << fruit->Consume() << endl;
    cout << vegetable->CookAndConsume() << endl;
    cout << vegetable->Serve() << endl;

    delete fruit;
    delete vegetable;
}

int main() {
    FirstFactory *f1 = new FirstFactory();
    Usage(*f1);
    delete f1;

    SecondFactory *f2 = new SecondFactory();
    Usage(*f2);
    delete f1;
}

File: abstractfactory/swift/myabstractfactory.swift
protocol Fruit {
    func consume() -> String
}

struct Apple: Fruit {
    func consume() -> String {
        return "Cut and Eat."
    }
}

struct Banana: Fruit {
    func consume() -> String {
        return "Peel and Eat."
    }
}

protocol Vegetable {
    func cookandconsume() -> String
    func serve() -> String
}

struct Potato: Vegetable {
    func cookandconsume() -> String {
        return "Fry and Eat"
    }

    func serve() -> String {
        return "In a bag"
    }
}

struct Beans: Vegetable {
    func cookandconsume() -> String {
        return "Boil and Eat"
    }

    func serve() -> String {
        return "In a bowl"
    }
}

protocol Factory {
    func consumeFruit() -> Fruit
    func consumeVegetable() -> Vegetable
}

struct FruitFactory {
    func consumeFruit() -> Fruit {
        let f = Apple()
        return f
    }

    func consumeVegetable() -> Vegetable {
        let v = Potato()
        return v
    }
}

struct VegetableFactory {
    func consumeFruit() -> Fruit {
        let f = Banana()
        return f
    }

    func consumeVegetable() -> Vegetable {
        let v = Beans()
        return v
    }
}

let f1 = FruitFactory()
let f2 = VegetableFactory()

let a = f1.consumeFruit()
let b = f2.consumeFruit()

print(a.consume())
print(b.consume())

let p = f1.consumeVegetable()
let be = f2.consumeVegetable()

print(p.cookandconsume())
print(p.serve())

print(be.cookandconsume())
print(be.serve())
